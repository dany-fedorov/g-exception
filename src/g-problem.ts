import * as hbs from 'handlebars';

type GProblemInfo = Record<string, any>;

export class GProblem {
  private readonly message: string;
  private readonly info: GProblemInfo | null = null;
  private type: string | null = null;

  constructor(message: string, info?: GProblemInfo) {
    if (info) {
      this.info = info;
    }
    try {
      this.message = hbs.compile(message)(info);
    } catch (caught: unknown) {
      console.warn(`GProblem error compiling a message: ${caught}`);
      this.message = message;
    }
  }

  static new(message: string, info?: GProblemInfo): GProblem {
    return new GProblem(message, info);
  }

  getInfo(): GProblemInfo | null {
    return this.info;
  }

  getMessage(): string {
    return this.message;
  }

  setType(type: string): this {
    this.type = type;
    return this;
  }

  getType(): string | null {
    return this.type;
  }
}
