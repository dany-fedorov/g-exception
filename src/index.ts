export { GExceptionV0, GExceptionConstructorArguments } from './GExceptionV0';
