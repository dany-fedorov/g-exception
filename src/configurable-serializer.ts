import { GProblem } from './g-problem';

interface ConSConfig {
  maxDepth: number;
  format: ConSFormat;
}

export enum ConSFormat {
  JSON_V0 = 'jsonv0',
  // INLINE_V0 = 'inlinev0' // TODO
}

const CONS_DEFAULT_CONFIG: ConSConfig = {
  maxDepth: 20,
  format: ConSFormat.JSON_V0,
};

const ConS_FORMAT_KEY = '$conS$>format';
const ConS_CONTENT_KEY = '$conS$>content';

const ConS_TYPE_KEY = '$conS$>type';

const ConS_CONSTRUCTOR_NAME_KEY = '$conS$>constructorName';

enum ConSType {
  conSPlainObject = 'conSPlainObject',
  conSCustomObject = 'conSCustomObject',
  Date = 'Date',
  RegExp = 'RegExp',
  Buffer = 'Buffer',
  Error = 'Error',
  BigInt = 'BigInt',
  undefined = 'undefined',
  Map = 'Map',
  Set = 'Set',
  Function = 'Function',
  ArrayBuffer = 'ArrayBuffer',
  Int8Array = 'Int8Array',
  Uint8Array = 'Uint8Array',
  Uint8ClampedArray = 'Uint8ClampedArray',
  Int16Array = 'Int16Array',
  Uint16Array = 'Uint16Array',
  Int32Array = 'Int32Array',
  Uint32Array = 'Uint32Array',
  Float32Array = 'Float32Array',
  Float64Array = 'Float64Array',
  BigInt64Array = 'BigInt64Array',
  BigUint64Array = 'BigUint64Array',
}

interface ConSObject {
  [ConS_TYPE_KEY]: Omit<ConSType, ConSType.conSCustomObject>;
  [ConS_CONTENT_KEY]: ConSObjectContent;
}

interface ConSCustomObject {
  [ConS_TYPE_KEY]: ConSType.conSCustomObject;
  [ConS_CONSTRUCTOR_NAME_KEY]: string;
  [ConS_CONTENT_KEY]: ConSObjectContent;
}

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
type ConSContent =
  | string
  | number
  | boolean
  | null
  | ConSObject
  | ConSCustomObject
  | ConSArray;

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
type ConSObjectContent = Record<string, ConSContent>;

type ConSArray = Array<ConSContent>;

interface ConSSerializedJsonV0Conatiner {
  [ConS_FORMAT_KEY]: ConSFormat;
  [ConS_CONTENT_KEY]: ConSContent;
}

interface ConSSerializeResult<Serialized> {
  appliedConfig: ConSConfig;
  serialized: Serialized | null;
  problems: GProblem[];
}

function mkProblemReason(reason: string): string {
  return `${ConfigurableSerializer.name} problem: ${reason}`;
}

interface ConSJsonV0Context {
  depth: number;
  objId: number;
  objToObjId: WeakMap<object, string>;
}

function serializeObjectToJsonV0(
  obj: unknown,
  config: ConSConfig,
  ctx: ConSJsonV0Context,
): ConSContent {}

function serializeToJsonV0(
  obj: unknown,
  config: ConSConfig,
): ConSSerializeResult<ConSSerializedJsonV0Conatiner> {
  const ctx: ConSJsonV0Context = {
    depth: 0,
    objId: 0,
    objToObjId: new WeakMap(),
  };
  const { serializedObject, problems } = serializeObjectToJsonV0(
    obj,
    config,
    ctx,
  );
  return {
    appliedConfig: config,
    serialized: {
      [ConS_FORMAT_KEY]: ConSFormat.JSON_V0,
      [ConS_CONTENT_KEY]: serializedObject,
    },
    problems,
  };
}

export class ConfigurableSerializer {
  private readonly config: ConSConfig;

  constructor(config: ConSConfig = CONS_DEFAULT_CONFIG) {
    this.config = config;
  }

  serialize(obj: unknown): ConSSerializeResult<ConSSerializedJsonV0Conatiner> {
    switch (this.config.format) {
      case ConSFormat.JSON_V0:
        return serializeToJsonV0(obj, this.config);
      default:
        return {
          appliedConfig: this.config,
          serialized: null,
          problems: [
            new GProblem(mkProblemReason(`Unknown format - {{format}}`), {
              format: this.config.format,
            }),
          ],
        };
    }
  }
}
