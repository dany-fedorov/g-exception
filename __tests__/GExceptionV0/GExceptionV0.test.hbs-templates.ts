import { GExceptionV0 } from '../../src/GExceptionV0';
import {
  MOCK_DISPLAY_MESSAGE,
  MOCK_ERR_CODE,
  MOCK_ERR_ID,
  MOCK_ERR_MSG,
  MOCK_ERR_NUM_CODE,
} from '../lib/mock-values';

beforeAll(() => {
  GExceptionV0.mergeConfig({
    logConstructorProblems: false,
  });
});

afterAll(() => {
  GExceptionV0.mergeConfig({
    logConstructorProblems: true,
  });
});
const TEMPLATE = [
  `double braces / code:                 ${GExceptionV0.t.code}`,
  `double braces / numCode:              ${GExceptionV0.t.numCode}`,
  `double braces / id:                   ${GExceptionV0.t.id}`,
  `double braces / displayMessage:       ${GExceptionV0.t.displayMessage}`,
  `double braces / timestamp:            ${GExceptionV0.t.timestamp}`,
  `double braces / causes:               ${GExceptionV0.t.causes()}`,
  `double braces / causes.json:          ${GExceptionV0.t.causes.json}`,
  `double braces / cause0:               ${GExceptionV0.t.causes(0)}`,
  `double braces / cause1:               ${GExceptionV0.t.causes(1)}`,
  `double braces / info:                 ${GExceptionV0.t.info()}`,
  `double braces / info.json:            ${GExceptionV0.t.info.json}`,
  `double braces / info.prop1:           ${GExceptionV0.t.info('prop1')}`,
  `double braces / info.prop2.prop22:    ${GExceptionV0.t.info(
    'prop2.prop22',
  )}`,
  `double braces / info.prop3:           ${GExceptionV0.t.info('prop3')}`,
  `double braces / info.prop4.prop44[0]: ${GExceptionV0.t.info(
    'prop4.prop44.[0]',
  )}`,
  '---',
  `triple braces / code:                 ${GExceptionV0._tt.code}`,
  `triple braces / numCode:              ${GExceptionV0._tt.numCode}`,
  `triple braces / id:                   ${GExceptionV0._tt.id}`,
  `triple braces / displayMessage:       ${GExceptionV0._tt.displayMessage}`,
  `triple braces / timestamp:            ${GExceptionV0._tt.timestamp}`,
  `triple braces / causes:               ${GExceptionV0._tt.causes()}`,
  `triple braces / causes.json:          ${GExceptionV0._tt.causes.json}`,
  `triple braces / cause0:               ${GExceptionV0._tt.causes(0)}`,
  `triple braces / cause1:               ${GExceptionV0._tt.causes(1)}`,
  `triple braces / info:                 ${GExceptionV0._tt.info()}`,
  `triple braces / info.json:            ${GExceptionV0._tt.info.json}`,
  `triple braces / info.prop1:           ${GExceptionV0._tt.info('prop1')}`,
  `triple braces / info.prop2.prop22:    ${GExceptionV0._tt.info(
    'prop2.prop22',
  )}`,
  `triple braces / info.prop3:           ${GExceptionV0._tt.info('prop3')}`,
  `triple braces / info.prop4.prop44[0]: ${GExceptionV0._tt.info(
    'prop4.prop44.[0]',
  )}`,
  '---',
  '---',
  'double braces / recursive / message: BEGIN',
  GExceptionV0.t.message,
  'double braces / recursive / message: END',
  '---',
  '---',
  'triple braces / recursive / message: BEGIN',
  GExceptionV0._tt.message,
  'triple braces / recursive / message: END',
].join('\n');

const configure = (e: GExceptionV0): GExceptionV0 => {
  return e
    .setCode(MOCK_ERR_CODE)
    .setNumCode(MOCK_ERR_NUM_CODE)
    .setId(MOCK_ERR_ID)
    .setDisplayMessage(MOCK_DISPLAY_MESSAGE)
    .setTimestamp(new Date('2022-07-13T11:12:25.741Z').toISOString())
    .setCauses([1, 2, 3])
    .setInfo({
      prop1: 'value of info.prop1',
      prop2: { prop22: 'value of info.prop2.prop22' },
    })
    .setInfoProp('prop3', 'value of info.prop3')
    .setInfoProp('prop4', { prop44: ['value of info.prop4.prop44[0]'] });
};

function removeStackEntries(s: string | undefined): string | undefined {
  if (!s) {
    return s;
  }
  return s.replace(/(\n\s*at).+/g, '$1');
}

describe('Message templating', () => {
  test('Valid', () => {
    const message = `<message>\n${TEMPLATE}\n</message>`;
    const e1 = configure(GExceptionV0.new(message));
    expect(e1.getMessage()).toMatchSnapshot('message:message');
    expect(removeStackEntries(e1.getStack())).toMatchSnapshot('message:stack');
    const displayMessage = `<displayMessage>\n${TEMPLATE}\n</displayMessage>`;
    const e2 = configure(GExceptionV0.new(MOCK_ERR_MSG)).setDisplayMessage(
      displayMessage,
    );
    expect(e2.getDisplayMessage()).toMatchSnapshot(
      'displayMessage:displayMessage',
    );
    expect(removeStackEntries(e2.getStack())).toMatchSnapshot(
      'displayMessage:stack',
    );
  });

  test('Invalid template', () => {
    GExceptionV0.from({ message: 'heh{{h}' });
  });
});
