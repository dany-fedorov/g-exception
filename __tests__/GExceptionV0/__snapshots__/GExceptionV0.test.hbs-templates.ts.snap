// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message templating Invalid template: stack 1`] = `
"Error: heh{{h}
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at"
`;

exports[`Message templating Invalid template: stack 2`] = `
"Error: heh{{h}
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at"
`;

exports[`Message templating Invalid template: stack 3`] = `
"Error: heh{{h}
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at"
`;

exports[`Message templating Invalid template: stack 4`] = `
Object {
  "handlebarsProblems": Array [
    Object {
      "caught": [Error: Parse error on line 1:
Error: heh{{h}    at Function.fr
-------------^
Expecting 'CLOSE_RAW_BLOCK', 'CLOSE', 'CLOSE_UNESCAPED', 'OPEN_SEXPR', 'CLOSE_SEXPR', 'ID', 'OPEN_BLOCK_PARAMS', 'STRING', 'NUMBER', 'BOOLEAN', 'UNDEFINED', 'NULL', 'DATA', 'SEP', got 'INVALID'],
      "problemReason": "GException problem: Problem during handlebars compilation of template \\"stack\\"
- template: Error: heh{{h}
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
- caught: Error: Parse error on line 1:
Error: heh{{h}    at Function.fr
-------------^
Expecting 'CLOSE_RAW_BLOCK', 'CLOSE', 'CLOSE_UNESCAPED', 'OPEN_SEXPR', 'CLOSE_SEXPR', 'ID', 'OPEN_BLOCK_PARAMS', 'STRING', 'NUMBER', 'BOOLEAN', 'UNDEFINED', 'NULL', 'DATA', 'SEP', got 'INVALID'",
      "template": "Error: heh{{h}
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at",
      "templatePropName": "stack",
    },
  ],
}
`;

exports[`Message templating Valid: displayMessage:displayMessage 1`] = `
"<displayMessage>
double braces / code:                 MOCK_ERR_CODE
double braces / numCode:              123
double braces / id:                   MOCK_ERR_ID
double braces / displayMessage:       &lt;displayMessage&gt;
double braces / code:                 {{code}}
double braces / numCode:              {{numCode}}
double braces / id:                   {{id}}
double braces / displayMessage:       {{displayMessage}}
double braces / timestamp:            {{timestamp}}
double braces / causes:               {{causes}}
double braces / causes.json:          {{json causes 0}}
double braces / cause0:               {{causes.[0]}}
double braces / cause1:               {{causes.[1]}}
double braces / info:                 {{info}}
double braces / info.json:            {{json info 0}}
double braces / info.prop1:           {{info.prop1}}
double braces / info.prop2.prop22:    {{info.prop2.prop22}}
double braces / info.prop3:           {{info.prop3}}
double braces / info.prop4.prop44[0]: {{info.prop4.prop44.[0]}}
---
triple braces / code:                 {{{code}}}
triple braces / numCode:              {{{numCode}}}
triple braces / id:                   {{{id}}}
triple braces / displayMessage:       {{{displayMessage}}}
triple braces / timestamp:            {{{timestamp}}}
triple braces / causes:               {{{causes}}}
triple braces / causes.json:          {{{json causes 0}}}
triple braces / cause0:               {{{causes.[0]}}}
triple braces / cause1:               {{{causes.[1]}}}
triple braces / info:                 {{{info}}}
triple braces / info.json:            {{{json info 0}}}
triple braces / info.prop1:           {{{info.prop1}}}
triple braces / info.prop2.prop22:    {{{info.prop2.prop22}}}
triple braces / info.prop3:           {{{info.prop3}}}
triple braces / info.prop4.prop44[0]: {{{info.prop4.prop44.[0]}}}
---
---
double braces / recursive / message: BEGIN
{{message}}
double braces / recursive / message: END
---
---
triple braces / recursive / message: BEGIN
{{{message}}}
triple braces / recursive / message: END
&lt;/displayMessage&gt;
double braces / timestamp:            2022-07-13T11:12:25.741Z
double braces / causes:               1,2,3
double braces / causes.json:          [1,2,3]
double braces / cause0:               1
double braces / cause1:               2
double braces / info:                 [object Object]
double braces / info.json:            {&quot;prop1&quot;:&quot;value of info.prop1&quot;,&quot;prop2&quot;:{&quot;prop22&quot;:&quot;value of info.prop2.prop22&quot;},&quot;prop3&quot;:&quot;value of info.prop3&quot;,&quot;prop4&quot;:{&quot;prop44&quot;:[&quot;value of info.prop4.prop44[0]&quot;]}}
double braces / info.prop1:           value of info.prop1
double braces / info.prop2.prop22:    value of info.prop2.prop22
double braces / info.prop3:           value of info.prop3
double braces / info.prop4.prop44[0]: value of info.prop4.prop44[0]
---
triple braces / code:                 MOCK_ERR_CODE
triple braces / numCode:              123
triple braces / id:                   MOCK_ERR_ID
triple braces / displayMessage:       <displayMessage>
double braces / code:                 {{code}}
double braces / numCode:              {{numCode}}
double braces / id:                   {{id}}
double braces / displayMessage:       {{displayMessage}}
double braces / timestamp:            {{timestamp}}
double braces / causes:               {{causes}}
double braces / causes.json:          {{json causes 0}}
double braces / cause0:               {{causes.[0]}}
double braces / cause1:               {{causes.[1]}}
double braces / info:                 {{info}}
double braces / info.json:            {{json info 0}}
double braces / info.prop1:           {{info.prop1}}
double braces / info.prop2.prop22:    {{info.prop2.prop22}}
double braces / info.prop3:           {{info.prop3}}
double braces / info.prop4.prop44[0]: {{info.prop4.prop44.[0]}}
---
triple braces / code:                 {{{code}}}
triple braces / numCode:              {{{numCode}}}
triple braces / id:                   {{{id}}}
triple braces / displayMessage:       {{{displayMessage}}}
triple braces / timestamp:            {{{timestamp}}}
triple braces / causes:               {{{causes}}}
triple braces / causes.json:          {{{json causes 0}}}
triple braces / cause0:               {{{causes.[0]}}}
triple braces / cause1:               {{{causes.[1]}}}
triple braces / info:                 {{{info}}}
triple braces / info.json:            {{{json info 0}}}
triple braces / info.prop1:           {{{info.prop1}}}
triple braces / info.prop2.prop22:    {{{info.prop2.prop22}}}
triple braces / info.prop3:           {{{info.prop3}}}
triple braces / info.prop4.prop44[0]: {{{info.prop4.prop44.[0]}}}
---
---
double braces / recursive / message: BEGIN
{{message}}
double braces / recursive / message: END
---
---
triple braces / recursive / message: BEGIN
{{{message}}}
triple braces / recursive / message: END
</displayMessage>
triple braces / timestamp:            2022-07-13T11:12:25.741Z
triple braces / causes:               1,2,3
triple braces / causes.json:          [1,2,3]
triple braces / cause0:               1
triple braces / cause1:               2
triple braces / info:                 [object Object]
triple braces / info.json:            {\\"prop1\\":\\"value of info.prop1\\",\\"prop2\\":{\\"prop22\\":\\"value of info.prop2.prop22\\"},\\"prop3\\":\\"value of info.prop3\\",\\"prop4\\":{\\"prop44\\":[\\"value of info.prop4.prop44[0]\\"]}}
triple braces / info.prop1:           value of info.prop1
triple braces / info.prop2.prop22:    value of info.prop2.prop22
triple braces / info.prop3:           value of info.prop3
triple braces / info.prop4.prop44[0]: value of info.prop4.prop44[0]
---
---
double braces / recursive / message: BEGIN
MOCK_ERR_MSG
double braces / recursive / message: END
---
---
triple braces / recursive / message: BEGIN
MOCK_ERR_MSG
triple braces / recursive / message: END
</displayMessage>"
`;

exports[`Message templating Valid: displayMessage:stack 1`] = `
"Error: MOCK_ERR_MSG
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at"
`;

exports[`Message templating Valid: message:message 1`] = `
"<message>
double braces / code:                 MOCK_ERR_CODE
double braces / numCode:              123
double braces / id:                   MOCK_ERR_ID
double braces / displayMessage:       MOCK_DISPLAY_MESSAGE
double braces / timestamp:            2022-07-13T11:12:25.741Z
double braces / causes:               1,2,3
double braces / causes.json:          [1,2,3]
double braces / cause0:               1
double braces / cause1:               2
double braces / info:                 [object Object]
double braces / info.json:            {&quot;prop1&quot;:&quot;value of info.prop1&quot;,&quot;prop2&quot;:{&quot;prop22&quot;:&quot;value of info.prop2.prop22&quot;},&quot;prop3&quot;:&quot;value of info.prop3&quot;,&quot;prop4&quot;:{&quot;prop44&quot;:[&quot;value of info.prop4.prop44[0]&quot;]}}
double braces / info.prop1:           value of info.prop1
double braces / info.prop2.prop22:    value of info.prop2.prop22
double braces / info.prop3:           value of info.prop3
double braces / info.prop4.prop44[0]: value of info.prop4.prop44[0]
---
triple braces / code:                 MOCK_ERR_CODE
triple braces / numCode:              123
triple braces / id:                   MOCK_ERR_ID
triple braces / displayMessage:       MOCK_DISPLAY_MESSAGE
triple braces / timestamp:            2022-07-13T11:12:25.741Z
triple braces / causes:               1,2,3
triple braces / causes.json:          [1,2,3]
triple braces / cause0:               1
triple braces / cause1:               2
triple braces / info:                 [object Object]
triple braces / info.json:            {\\"prop1\\":\\"value of info.prop1\\",\\"prop2\\":{\\"prop22\\":\\"value of info.prop2.prop22\\"},\\"prop3\\":\\"value of info.prop3\\",\\"prop4\\":{\\"prop44\\":[\\"value of info.prop4.prop44[0]\\"]}}
triple braces / info.prop1:           value of info.prop1
triple braces / info.prop2.prop22:    value of info.prop2.prop22
triple braces / info.prop3:           value of info.prop3
triple braces / info.prop4.prop44[0]: value of info.prop4.prop44[0]
---
---
double braces / recursive / message: BEGIN
&lt;message&gt;
double braces / code:                 {{code}}
double braces / numCode:              {{numCode}}
double braces / id:                   {{id}}
double braces / displayMessage:       {{displayMessage}}
double braces / timestamp:            {{timestamp}}
double braces / causes:               {{causes}}
double braces / causes.json:          {{json causes 0}}
double braces / cause0:               {{causes.[0]}}
double braces / cause1:               {{causes.[1]}}
double braces / info:                 {{info}}
double braces / info.json:            {{json info 0}}
double braces / info.prop1:           {{info.prop1}}
double braces / info.prop2.prop22:    {{info.prop2.prop22}}
double braces / info.prop3:           {{info.prop3}}
double braces / info.prop4.prop44[0]: {{info.prop4.prop44.[0]}}
---
triple braces / code:                 {{{code}}}
triple braces / numCode:              {{{numCode}}}
triple braces / id:                   {{{id}}}
triple braces / displayMessage:       {{{displayMessage}}}
triple braces / timestamp:            {{{timestamp}}}
triple braces / causes:               {{{causes}}}
triple braces / causes.json:          {{{json causes 0}}}
triple braces / cause0:               {{{causes.[0]}}}
triple braces / cause1:               {{{causes.[1]}}}
triple braces / info:                 {{{info}}}
triple braces / info.json:            {{{json info 0}}}
triple braces / info.prop1:           {{{info.prop1}}}
triple braces / info.prop2.prop22:    {{{info.prop2.prop22}}}
triple braces / info.prop3:           {{{info.prop3}}}
triple braces / info.prop4.prop44[0]: {{{info.prop4.prop44.[0]}}}
---
---
double braces / recursive / message: BEGIN
{{message}}
double braces / recursive / message: END
---
---
triple braces / recursive / message: BEGIN
{{{message}}}
triple braces / recursive / message: END
&lt;/message&gt;
double braces / recursive / message: END
---
---
triple braces / recursive / message: BEGIN
<message>
double braces / code:                 {{code}}
double braces / numCode:              {{numCode}}
double braces / id:                   {{id}}
double braces / displayMessage:       {{displayMessage}}
double braces / timestamp:            {{timestamp}}
double braces / causes:               {{causes}}
double braces / causes.json:          {{json causes 0}}
double braces / cause0:               {{causes.[0]}}
double braces / cause1:               {{causes.[1]}}
double braces / info:                 {{info}}
double braces / info.json:            {{json info 0}}
double braces / info.prop1:           {{info.prop1}}
double braces / info.prop2.prop22:    {{info.prop2.prop22}}
double braces / info.prop3:           {{info.prop3}}
double braces / info.prop4.prop44[0]: {{info.prop4.prop44.[0]}}
---
triple braces / code:                 {{{code}}}
triple braces / numCode:              {{{numCode}}}
triple braces / id:                   {{{id}}}
triple braces / displayMessage:       {{{displayMessage}}}
triple braces / timestamp:            {{{timestamp}}}
triple braces / causes:               {{{causes}}}
triple braces / causes.json:          {{{json causes 0}}}
triple braces / cause0:               {{{causes.[0]}}}
triple braces / cause1:               {{{causes.[1]}}}
triple braces / info:                 {{{info}}}
triple braces / info.json:            {{{json info 0}}}
triple braces / info.prop1:           {{{info.prop1}}}
triple braces / info.prop2.prop22:    {{{info.prop2.prop22}}}
triple braces / info.prop3:           {{{info.prop3}}}
triple braces / info.prop4.prop44[0]: {{{info.prop4.prop44.[0]}}}
---
---
double braces / recursive / message: BEGIN
{{message}}
double braces / recursive / message: END
---
---
triple braces / recursive / message: BEGIN
{{{message}}}
triple braces / recursive / message: END
</message>
triple braces / recursive / message: END
</message>"
`;

exports[`Message templating Valid: message:stack 1`] = `
"Error: <message>
double braces / code:                 MOCK_ERR_CODE
double braces / numCode:              123
double braces / id:                   MOCK_ERR_ID
double braces / displayMessage:       MOCK_DISPLAY_MESSAGE
double braces / timestamp:            2022-07-13T11:12:25.741Z
double braces / causes:               1,2,3
double braces / causes.json:          [1,2,3]
double braces / cause0:               1
double braces / cause1:               2
double braces / info:                 [object Object]
double braces / info.json:            {&quot;prop1&quot;:&quot;value of info.prop1&quot;,&quot;prop2&quot;:{&quot;prop22&quot;:&quot;value of info.prop2.prop22&quot;},&quot;prop3&quot;:&quot;value of info.prop3&quot;,&quot;prop4&quot;:{&quot;prop44&quot;:[&quot;value of info.prop4.prop44[0]&quot;]}}
double braces / info.prop1:           value of info.prop1
double braces / info.prop2.prop22:    value of info.prop2.prop22
double braces / info.prop3:           value of info.prop3
double braces / info.prop4.prop44[0]: value of info.prop4.prop44[0]
---
triple braces / code:                 MOCK_ERR_CODE
triple braces / numCode:              123
triple braces / id:                   MOCK_ERR_ID
triple braces / displayMessage:       MOCK_DISPLAY_MESSAGE
triple braces / timestamp:            2022-07-13T11:12:25.741Z
triple braces / causes:               1,2,3
triple braces / causes.json:          [1,2,3]
triple braces / cause0:               1
triple braces / cause1:               2
triple braces / info:                 [object Object]
triple braces / info.json:            {\\"prop1\\":\\"value of info.prop1\\",\\"prop2\\":{\\"prop22\\":\\"value of info.prop2.prop22\\"},\\"prop3\\":\\"value of info.prop3\\",\\"prop4\\":{\\"prop44\\":[\\"value of info.prop4.prop44[0]\\"]}}
triple braces / info.prop1:           value of info.prop1
triple braces / info.prop2.prop22:    value of info.prop2.prop22
triple braces / info.prop3:           value of info.prop3
triple braces / info.prop4.prop44[0]: value of info.prop4.prop44[0]
---
---
double braces / recursive / message: BEGIN
&lt;message&gt;
double braces / code:                 {{code}}
double braces / numCode:              {{numCode}}
double braces / id:                   {{id}}
double braces / displayMessage:       {{displayMessage}}
double braces / timestamp:            {{timestamp}}
double braces / causes:               {{causes}}
double braces / causes.json:          {{json causes 0}}
double braces / cause0:               {{causes.[0]}}
double braces / cause1:               {{causes.[1]}}
double braces / info:                 {{info}}
double braces / info.json:            {{json info 0}}
double braces / info.prop1:           {{info.prop1}}
double braces / info.prop2.prop22:    {{info.prop2.prop22}}
double braces / info.prop3:           {{info.prop3}}
double braces / info.prop4.prop44[0]: {{info.prop4.prop44.[0]}}
---
triple braces / code:                 {{{code}}}
triple braces / numCode:              {{{numCode}}}
triple braces / id:                   {{{id}}}
triple braces / displayMessage:       {{{displayMessage}}}
triple braces / timestamp:            {{{timestamp}}}
triple braces / causes:               {{{causes}}}
triple braces / causes.json:          {{{json causes 0}}}
triple braces / cause0:               {{{causes.[0]}}}
triple braces / cause1:               {{{causes.[1]}}}
triple braces / info:                 {{{info}}}
triple braces / info.json:            {{{json info 0}}}
triple braces / info.prop1:           {{{info.prop1}}}
triple braces / info.prop2.prop22:    {{{info.prop2.prop22}}}
triple braces / info.prop3:           {{{info.prop3}}}
triple braces / info.prop4.prop44[0]: {{{info.prop4.prop44.[0]}}}
---
---
double braces / recursive / message: BEGIN
{{message}}
double braces / recursive / message: END
---
---
triple braces / recursive / message: BEGIN
{{{message}}}
triple braces / recursive / message: END
&lt;/message&gt;
double braces / recursive / message: END
---
---
triple braces / recursive / message: BEGIN
<message>
double braces / code:                 {{code}}
double braces / numCode:              {{numCode}}
double braces / id:                   {{id}}
double braces / displayMessage:       {{displayMessage}}
double braces / timestamp:            {{timestamp}}
double braces / causes:               {{causes}}
double braces / causes.json:          {{json causes 0}}
double braces / cause0:               {{causes.[0]}}
double braces / cause1:               {{causes.[1]}}
double braces / info:                 {{info}}
double braces / info.json:            {{json info 0}}
double braces / info.prop1:           {{info.prop1}}
double braces / info.prop2.prop22:    {{info.prop2.prop22}}
double braces / info.prop3:           {{info.prop3}}
double braces / info.prop4.prop44[0]: {{info.prop4.prop44.[0]}}
---
triple braces / code:                 {{{code}}}
triple braces / numCode:              {{{numCode}}}
triple braces / id:                   {{{id}}}
triple braces / displayMessage:       {{{displayMessage}}}
triple braces / timestamp:            {{{timestamp}}}
triple braces / causes:               {{{causes}}}
triple braces / causes.json:          {{{json causes 0}}}
triple braces / cause0:               {{{causes.[0]}}}
triple braces / cause1:               {{{causes.[1]}}}
triple braces / info:                 {{{info}}}
triple braces / info.json:            {{{json info 0}}}
triple braces / info.prop1:           {{{info.prop1}}}
triple braces / info.prop2.prop22:    {{{info.prop2.prop22}}}
triple braces / info.prop3:           {{{info.prop3}}}
triple braces / info.prop4.prop44[0]: {{{info.prop4.prop44.[0]}}}
---
---
double braces / recursive / message: BEGIN
{{message}}
double braces / recursive / message: END
---
---
triple braces / recursive / message: BEGIN
{{{message}}}
triple braces / recursive / message: END
</message>
triple braces / recursive / message: END
</message>
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at
    at"
`;
